### PR TITLE
feat: add filtered CSV export for issues

### LINKED ISSUE #82: Export issues to CSV
We import from Jira (`backend/lib_softtrack/importer.py`) but offer no export of any kind. Self-hosted users need a way out as much as a way in — for backups, spreadsheets, or leaving for another tool. The README roadmap already promises CSV export; this issue makes it concrete.

**Scope**
- `GET /teams/{team_id}/issues/export` returning `text/csv` as a streaming response
- Respects the same filters as the list endpoint (`backend/app_softtrack/issues.py`) so "export what I'm looking at" works naturally
- Columns: key (`ENG-42`), title, description, status, priority, assignee, labels (semicolon-joined), project, cycle, estimate, creator, created, updated, parent key
- UTF-8 with BOM so Excel opens it correctly; RFC 4180 quoting (Python's `csv` module handles this)
- Frontend: an "Export CSV" action in the board/list overflow menu that hits the endpoint with current filters

**Out of scope**: JSON export, attachments, comments — follow-ups if anyone wants them.

Tests should cover the quoting edge cases (titles with commas, quotes, newlines) — see `backend/tests/` for the existing service-test style.

### PR DESCRIPTION
**Pull request by LubabaKhalid**
Closes #82 
## What this changes

Adds filtered CSV export for issues, including readable created/updated timestamps and an **Export CSV** button in the issue board. The export respects the currently selected issue filters.

## Why

The issue board did not provide a way to export the currently filtered issues. This adds an unpaginated CSV export using the same filtering logic as the issue list, so the exported data matches the selected filters.

## How it was verified

* Manually tested CSV export from the issue board.
* Verified status, priority, and combined filters produce the expected issues.
* Verified the CSV contains the expected issue fields and readable timestamps (`YYYY-MM-DD HH:MM:SS`).
* Backend test suite: **653 passed, 0 failed, 0 skipped**.
* CSV export tests: **5 passed**.
* Frontend TypeScript check passed.
* Frontend production build passed.
* Docker frontend build passed.

## Checklist

* [x] `cd backend && black . && pytest` passes
* [ ] If backend routes or schemas changed, the API client was regenerated and committed (`npm run export:openapi && npm run generate:api` in `frontend/`)
* [x] If the schema changed, there is an Alembic migration and I have read what autogenerate wrote — **No schema change**
* [x] New behaviour has a test, and I checked that the test fails without the change
* [x] Business logic went in `lib_softtrack/`, not in a route handler

### CODE REVIEW

**Comment by LubabaKhalid** [comment]
@wak327 Backend, API contract, and Docker checks are passing. The only failing check is the frontend test due to "sessionStorage is not defined" in "frontend/src/auth/__tests__/oauth.test.ts", which is from the upstream merge commit and not present in my branch.

Could you please take a look at the frontend test issue and review the PR when possible?

**Comment by wak327** [comment]
sync the fork, and than update your branch with main

**Comment by LubabaKhalid** [comment]
Hi, I’ve synced my fork with  main and updated the branch with the latest changes. The changes are pushed to PR #165 and all CI checks are passing. Thanks

File: backend/app_softtrack/issues.py [hunk 1]
```diff
@@ -65,6 +81,122 @@ def list_issues(
...
+
+            labels = (
+                ";".join([label.name for label in issue.labels]) if issue.labels else ""
+            )
+
+            project_name = ""
+            if issue.project_id is not None:
+                from lib_softtrack.tables import Project
+
+                proj = session.get(Project, issue.project_id)
```
Comment (wak327) [inline backend/app_softtrack/issues.py:157]: **Blocking — these two `session.get()` calls run after the request session has closed.**

`iter_rows` is iterated by Starlette *after* `export_issues_csv` returns. In FastAPI >= 0.106 the `yield`-dependency exit stack closes when the endpoint function returns, before the response body is consumed — so by the time line 157 (and 165) execute, the `with Session(engine)` block in `web.get_session` has already exited.

SQLAlchemy won't raise here: `close()` isn't `dispose()`, so each `get()` silently checks out a *new* connection outside the request lifecycle, with no deterministic close, and no way to turn a failure into an HTTP status because the headers are already sent.

The suite can't catch this. `conftest.client_fixture` overrides the dependency with `lambda: session` — a plain function, not a generator — so there is no teardown at all and the session stays open for the whole test.

Everything else the generator touches (`issue.assignee`, `.labels`, `.parent`, `.creator`, `.status`) is already-materialized `IssueRead` data, so the fix is small: resolve the names *before* building the response.

```python
from lib_softtrack.tables import Cycle, Project  # top of file

project_names = {
    p.id: p.name
    for p in session.exec(
        select(Project).where(
            Project.id.in_({i.project_id for i in issues if i.project_id})
        )
    ).all()
}
cycle_names = {
    c.id: c.name or f"Cycle {c.number}"
    for c in session.exec(
        select(Cycle).where(Cycle.id.in_({i.cycle_id for i in issues if i.cycle_id}))
    ).all()
}
```

That also removes the per-row queries, the imports inside the loop body, and the duplicated cycle-name idiom below.

File: backend/lib_softtrack/issues.py [hunk 2]
```diff
@@ -316,25 +364,45 @@ def list_issues(
...
+        unassigned=unassigned,
+        label_id=label_id,
+        parent_id=parent_id,
+        cycle_id=cycle_id,
     )
 
+    issues = session.exec(
+        select(Issue).where(*filters).order_by(Issue.number.desc())
+    ).all()
+    return _expand_issues(list(issues), session)
```
Comment (wak327) [inline backend/lib_softtrack/issues.py:404]: **Blocking — nothing is actually streamed, and the export is unbounded.**

`export_issues` materializes every matching issue through `_expand_issues` before `StreamingResponse` is even constructed, so peak memory is the full result set either way. The streaming machinery buys nothing here while costing us the closed-session problem in `app_softtrack/issues.py`.

More importantly, this defeats the property `_expand_issues` was written for. Its docstring promises "a fixed number of queries... constant in the page size", which it achieves with `WHERE id IN (:ids)`. Called unpaginated, that becomes one bind parameter per issue — Postgres caps a statement at 65535 parameters, and the IN-list is pathological long before that. The largest export any test exercises is 61 issues.

Two workable directions:
- Chunk the fetch — batch ids in groups of ~500 through `_expand_issues` and yield each batch's rows as you go. This also makes the streaming real, so the response class stops being a lie.
- Or put a documented cap on the export and return a clear error past it.

Either is fine; silently OOM-ing or 500-ing on a large team is the outcome to avoid.

File: backend/tests/test_issues.py [hunk 3]
```diff
@@ -181,3 +181,96 @@ def test_deleting_a_bare_issue(client, issue, team):
...
+        "estimate",
+        "creator",
+        "created",
+        "updated",
+        "parent_key",
+    ]
+
+    # Find our special issue row by title
+    titles = [r[1] for r in rows[1:]]
+    assert any('Title, with comma "quote"' in t for t in titles)
```
Comment (wak327) [inline backend/tests/test_issues.py:228]: **Blocking — six of the fourteen columns are only ever asserted as empty strings.**

Every issue created across the three export tests has no project, cycle, labels, assignee, estimate, or parent. This test and `test_export_respects_filters_and_is_unpaginated` build title/description/priority only, and the `issue` fixture is a bare title. So `project`, `cycle`, `labels`, `assignee`, `estimate`, and `parent_key` are never exercised with a real value — and the two `session.get()` branches have **zero** coverage.

The checklist says "New behaviour has a test, and I checked that the test fails without the change", but most of the row-building logic would survive being deleted outright.

The `board` fixture in `test_issue_filters.py` already gives you labels and assignees. One test that exports from `board` and asserts a fully-populated row — project, cycle, `;`-joined labels, assignee, estimate, parent key — would close most of this gap.

Smaller note on the assertion at line 228: `any('Title, with comma "quote"' in t for t in titles)` passes if *any* row happens to contain that substring. Asserting the exact round-tripped title is stronger and just as short.

File: backend/tests/conftest.py [hunk 4]
```diff
@@ -30,8 +30,30 @@
...
+    # Determine the dialect from the SQLAlchemy connection record.
+    try:
+        conn = getattr(connection_record, "driver_connection", None)
+        engine = getattr(conn, "engine", None)
+        if engine is not None:
+            if engine.dialect.name != "sqlite":
+                return
+    except Exception:
+        # If anything goes wrong, fall back to DBAPI check below.
+        pass
```
Comment (wak327) [inline backend/tests/conftest.py:50]: **This change is unrelated to CSV export, and the block it adds never executes.**

Two separate problems.

**It's dead code.** `connection_record.driver_connection` is the raw `sqlite3.Connection`, which has no `.engine` attribute — so `engine` is always `None`, `if engine is not None` never fires, and control always falls through to the `__module__` check below. Probed against the pinned SQLAlchemy 2.0.45:

```
{'record_type': '_ConnectionRecord', 'driver_connection_type': 'Connection',
 'driver_conn_has_engine': False, 'dbapi_module': 'sqlite3'}
```

The `except Exception: pass` is what hides this — with a bare `getattr` chain there's nothing to catch, so the handler only serves to make a permanently-broken branch look load-bearing. Only the two-line fallback at 52-55 does any work.

**And it's solving a problem we don't have.** The module docstring says every test gets in-memory SQLite, and CI pins `DATABASE_URL: "sqlite://"` with the comment "Tests use in-memory SQLite" (`.github/workflows/ci.yml`). There is no Postgres connection for this listener to no-op against.

I'd revert the whole hunk. If a Postgres test path is actually planned, that's worth its own PR where the guard can be tested.

File: frontend/src/board/TopBar.tsx [hunk 5]
```diff
@@ -119,6 +122,8 @@ export function TopBar({
           onClose={onCloseNotifications}
         />
 
+        <ExportCsvButton filters={filters} />
```
Comment (wak327) [inline frontend/src/board/TopBar.tsx:125]: **The export ignores the active search query, so it can hand back something quite different from what's on screen.**

`toQueryParams` carries no search term — search goes through a separate endpoint. While `searchQuery` is set, `BoardPage` replaces the entire list with `<SearchResults>`, but `TopBar` and this button stay rendered and enabled. So someone looking at 3 search hits clicks Export CSV, gets all 1,200 filtered issues, and has no indication why.

The PR description's "respects the currently selected issue filters" is accurate but incomplete in a way a user would read as a bug.

`BoardPage` already has the convention for this one line below where it renders `TopBar`:

```tsx
{selectedCycle && !searchQuery && <CycleBanner cycle={selectedCycle} />}
```

Hiding or disabling the button while a search is active would match it. Disabled with a `title` explaining why is probably kinder than it vanishing.

File: backend/lib_softtrack/issues.py [hunk 6]
```diff
@@ -296,14 +348,10 @@ def list_issues(
...
         filters.append(Issue.assignee_id == None)  # noqa: E711 -- SQL IS NULL
     elif assignee_id is not None:
         filters.append(Issue.assignee_id == assignee_id)
     if label_id is not None:
-        # A subquery rather than a join: an issue joined to its label links
-        # would come back once per matching link, and the count above would
-        # count it that many times.
         filters.append(
             Issue.id.in_(
                 select(IssueLabelLink.issue_id).where(
```
Comment (wak327) [inline backend/lib_softtrack/issues.py:357]: **Three rationale comments were dropped rather than moved with the code.**

The refactor lost:

```python
# Distinct from leaving assignee_id null, which means "anybody".
```

and, above the `label_id` subquery:

```python
# A subquery rather than a join: an issue joined to its label links
# would come back once per matching link, and the count above would
# count it that many times.
```

This codebase documents *why* pretty consistently, and the subquery one is specifically the kind of thing someone "simplifies" into a join in six months — at which point `total` starts double-counting and nothing fails loudly. Worth carrying both across verbatim.

The count it refers to now lives in `list_issues` rather than directly above, so the wording needs a small tweak ("the count in `list_issues`").

File: frontend/src/board/TopBar.tsx [hunk 7]
```diff
@@ -127,3 +132,44 @@ export function TopBar({
     </header>
   )
 }
+
+function ExportCsvButton({ filters }: { filters: any }) {
```
Comment (wak327) [inline frontend/src/board/TopBar.tsx:136]: `BoardFilters` is imported at the top of this very file, and `toQueryParams(filters: BoardFilters)` exists precisely to make this mapping typed — passing `any` into it discards the guarantee.

```tsx
function ExportCsvButton({ filters }: { filters: BoardFilters }) {
```

`oxlint` has `no-explicit-any` off so CI won't flag it, but with the right type in scope there's no cost to using it.

Separately: this component might read better in its own file under `frontend/src/board/`, given `TopBar.tsx` is otherwise pure layout. Take it or leave it.

File: frontend/src/board/TopBar.tsx [hunk 8]
```diff
@@ -127,3 +132,44 @@ export function TopBar({
...
+      const objectUrl = URL.createObjectURL(resp.data)
+      const link = document.createElement('a')
+      link.href = objectUrl
+      link.download = 'issues.csv'
+      document.body.appendChild(link)
+      link.click()
+      link.remove()
+    } finally {
+      setLoading(false)
+    }
```
Comment (wak327) [inline frontend/src/board/TopBar.tsx:160]: `try`/`finally` with no `catch` means a 403, a 500, or a dropped connection produces an unhandled promise rejection and no user feedback at all — the button just stops spinning and nothing downloads.

There's no toast system in the app yet, so local state is probably the pragmatic answer:

```tsx
} catch {
  setFailed(true)
} finally {
  setLoading(false)
}
```

...and render a short inline message. Compare `attachmentObjectUrl` in `attachments/urls.ts`, which handles its failure path deliberately and documents the reasoning.

File: frontend/src/board/TopBar.tsx [hunk 9]
```diff
@@ -127,3 +132,44 @@ export function TopBar({
...
+    setLoading(true)
+    try {
+      const qp = toQueryParams(filters)
+      const params = new URLSearchParams()
+      Object.entries(qp).forEach(([k, v]) => {
+        if (v !== undefined && v !== null) params.set(k, String(v))
+      })
+      const url = `/teams/${team.id}/issues/export?${params.toString()}`
+      const resp = await AXIOS_INSTANCE.get(url, { responseType: 'blob' })
+      const objectUrl = URL.createObjectURL(resp.data)
```
Comment (wak327) [inline frontend/src/board/TopBar.tsx:151]: The object URL is never revoked, so every click pins a full CSV blob for the lifetime of the page.

`downloadAttachment` also skips revocation, but it earns that: the URLs are cached by path and reused by live `` tags, and the docstring spells out that revoking would break them. Neither applies here — this URL is created fresh per click and used exactly once.

```tsx
link.click()
link.remove()
URL.revokeObjectURL(objectUrl)
```

File: backend/app_softtrack/issues.py [hunk 10]
```diff
@@ -65,6 +81,122 @@ def list_issues(
     )
 
 
+@router.get("/teams/{team_id}/issues/export")
```
Comment (wak327) [inline backend/app_softtrack/issues.py:84]: The route declares no `response_class` or `responses`, so FastAPI documents this `text/csv` endpoint as `application/json` with an empty schema (visible in the `openapi.json` diff). Anyone regenerating the client gets a JSON-typed hook for a CSV download.

```python
@router.get(
    "/teams/{team_id}/issues/export",
    response_class=StreamingResponse,
    responses={200: {"content": {"text/csv": {}}}},
)
```

File: backend/app_softtrack/issues.py [hunk 11]
```diff
@@ -65,6 +81,122 @@ def list_issues(
...
+                if proj is not None:
+                    project_name = proj.name
+
+            cycle_name = ""
+            if issue.cycle_id is not None:
+                from lib_softtrack.tables import Cycle
+
+                cyc = session.get(Cycle, issue.cycle_id)
+                if cyc is not None:
+                    cycle_name = cyc.name if cyc.name else f"Cycle {cyc.number}"
```
Comment (wak327) [inline backend/app_softtrack/issues.py:167]: This is the sixth copy of the `cycle.name or f"Cycle {cycle.number}"` idiom — `rules.py:339`, `cycles.py:107`, `cycles.py:204`, `reports.py:217`, `reports.py:272` — and the only one written as an `if`/`else` rather than an `or`.

`cycles.py:107` already calls the concept `display_name`. Worth extracting a small helper while you're in here, since the batched-lookup fix above touches this line anyway.

File: backend/app_softtrack/issues.py [hunk 12]
```diff
@@ -15,9 +16,24 @@
 from lib_softtrack.tables import IssuePriority, User
 from web import get_session
 
+import csv
+import io
+from datetime import datetime
+from typing import Iterable
```
Comment (wak327) [inline backend/app_softtrack/issues.py:22]: Two small import notes:

- These stdlib imports landed *after* the third-party and local blocks. CI only runs `black`, so nothing fails, but the file's existing block is ordered and this breaks it — they belong at the top.
- `from typing import Iterable` is deprecated in favour of `collections.abc.Iterable` (the `typing` aliases have been soft-deprecated since 3.9). The file already imports `Optional` from `typing`, which is fine to keep.

File: backend/openapi.json [hunk 13]
```diff
@@ -1873,6 +1873,184 @@
         }
       }
     },
+    "/teams/{team_id}/issues/export": {
```
Comment (wak327) [inline backend/openapi.json:1876]: `openapi.json` picked up the new endpoint but `frontend/src/api/generated/` did not — `git grep 'issues/export'` finds nothing under it. This is the unchecked checklist box, and it means the next person to run `npm run generate:api` gets an unrelated diff mixed into their PR.

Calling `AXIOS_INSTANCE` directly for a blob download is fine and has precedent (`attachments/urls.ts`, `ImportJiraModal.tsx`), so the generated hook doesn't need to be *used* — but the committed schema and the committed client should still land together. Worth doing after the `response_class` fix so the regenerated types are right.

**Review by wak327 — CHANGES_REQUESTED** [review]
Nice work on the core design — hoisting the filter logic into `build_issue_filters` so list and export can't drift apart is exactly the right call, and the CSV escaping, `\r\n` line terminator, BOM, and timestamp formatting are all handled correctly.

Requesting changes on three things before merge, plus a set of smaller items. The headline issue is that the two `session.get()` calls inside `iter_rows` execute after the request-scoped session has closed — see the inline comment on `app_softtrack/issues.py`. Fixing that one properly also resolves the per-row queries, the imports-in-the-loop, and the untested code paths, so it's a single tightly-scoped change.

**Blocking**
1. DB reads inside the streaming generator run after session teardown.
2. Nothing actually streams, and the export is unbounded.
3. Six of the fourteen CSV columns are never tested with a non-empty value.

**Should fix**
4. The `conftest.py` change is unrelated to this PR, and its main branch is dead code.
5. The export silently ignores the active search query.
6. Three explanatory comments were dropped in the refactor.
7. `filters: any`, no `catch`, and the object URL is never revoked.

**Minor** (inline where they anchor)
- `openapi.json` was regenerated but `frontend/src/api/generated/` was not.
- Missing `response_class`/`responses`, so a `text/csv` endpoint is documented as JSON.
- Sixth copy of the `cycle.name or f"Cycle {cycle.number}"` idiom.
- `build_issue_filters` is public-named but module-private (contrast `_expand_issues`) and annotated bare `-> list`.
- `_csv_timestamp(None)` is unreachable via `IssueRead`, whose `created_at`/`updated_at` are non-optional — so `test_export_timestamp_helper_handles_a_missing_value` covers a branch nothing can reach.
- No frontend test for the new button, despite the component-test setup added in #159.

One caveat on process: I couldn't execute the suite (no venv in `backend/`, `fastapi` not importable in my environment), so items 1-3 come from reading the code plus a SQLAlchemy probe, not a reproduced failure. Items 4 and 8 I verified directly. Worth confirming 1 against a real server rather than taking my word for it.

**Comment by LubabaKhalid** [comment]
Thanks for the detailed review. I went through all the points and addressed them.

The export now uses a separate session for the streaming generator, uses batched keyset pagination, and resolves the related data without per-row queries. I also added coverage for all CSV fields with real values.

I fixed the search handling, restored the explanatory comments, added proper error handling and URL cleanup on the frontend, regenerated the API client, fixed the CSV response documentation, and cleaned up the helper naming/types and cycle display-name handling.

For the frontend test item, I added `TopBar.test.tsx` with 9 tests covering filters, download handling, searching/disabled state, failures, and the export-in-progress state.

Verification:

* Backend: 751 tests passed
* Frontend: 203 tests passed
* TopBar tests: 9 passed
* TypeScript: passed
* Lint: passed with only 2 pre-existing warnings
* `git diff --check`: clean

I also used mutation checks on the main export/frontend behaviors to make sure the tests actually catch regressions.
